### PR TITLE
sqlcmd: init at 0.15.3

### PR DESCRIPTION
**Pull request by ratsclub**
###### Description of changes

Adds [go-sqlcmd](https://github.com/microsoft/go-sqlcmd).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by superherointj** [comment]
On order, please move:
  - vendorSha256 to after `src`.
  - subPackages to configure phase.
* To fix tests:
  - Error: mkdir /homeless-shelter: permission denied
  - On preCheck, use: `export HOME=$(mkdir -d)`
* Shell completion is missing. (bash, zsh, fish are available)
* Version (output) is wrong.
* Version test is missing.

**Comment by ratsclub** [comment]
@superherointj I didn't find the completion option. How did you find this out?

**Comment by superherointj** [comment]
> @superherointj I didn't find the completion option. How did you find this out?

./results/sqlcmd/bin/sqlcmd completion bash
./results/sqlcmd/bin/sqlcmd completion fish
./results/sqlcmd/bin/sqlcmd completion zsh

File: pkgs/development/tools/database/sqlcmd/default.nix [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    mv $out/bin/modern $out/bin/sqlcmd
+  '';
+
+  passthru = {
+    tests.version = testers.testVersion {
+      package = sqlcmd;
+      command = "sqlcmd --version";
+      inherit version;
+    };
+  };
```
Comment (superherointj) [inline pkgs/development/tools/database/sqlcmd/default.nix:41]:
```suggestion
  passthru.tests.version = testers.testVersion {
    inherit version;
    command = "sqlcmd --version";
    package = sqlcmd;    
  };
```

File: pkgs/development/tools/database/sqlcmd/default.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    sha256 = "sha256-1CdZVh7pbupCNOE1ydgYEqX4rYo2LRddlNRNf0QjlN0=";
+  };
+
+  vendorSha256 = "sha256-/51qs7ACE9UmyE59Tsto+Xm240H6IUL8eNKfLgb5LUE=";
+
+  ldflags = [ "-s" "-w" "-X main.version=${version}" ];
+
+  subPackages = [
+    "cmd/modern"
+  ];
```
Comment (superherointj) [inline pkgs/development/tools/database/sqlcmd/default.nix:25]:
```suggestion
  subPackages = [ "cmd/modern" ];
```

Comment (superherointj) [reply pkgs/development/tools/database/sqlcmd/default.nix:25]: As does not multiple items, one line is best.

File: pkgs/development/tools/database/sqlcmd/default.nix (hunk 1 above)
Comment (superherointj) [reply pkgs/development/tools/database/sqlcmd/default.nix:41]: As does not contain other items, one line is best.

**Comment by superherointj** [comment]
> Result of `nixpkgs-review pr 221699` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
> 1 package failed to build:
> 
> I have a hash mismatch when trying to build this
> 
> ```
> sqlcmd-0.15.3-go-modules> installing
> error: hash mismatch in fixed-output derivation '/nix/store/m8bpv64h5zvgl6w81pvw408k6h8nl48p-sqlcmd-0.15.3-go-modules.drv':
>          specified: sha256-/51qs7ACE9UmyE59Tsto+Xm240H6IUL8eNKfLgb5LUE=
>             got:    sha256-PuPP6BBentq99dUiJl6B4eD9avHzmbQOffP0vPdpdLo=
> error: 1 dependencies of derivation '/nix/store/x3psfrvx909pnvpmyflxz9kgbqmx8m55-sqlcmd-0.15.3.drv' failed to build
> error: 1 dependencies of derivation '/nix/store/rzrjxqchmpnrs90ra7qrp2m5cwias9if-review-shell.drv' failed to build
> ```

Thanks for reporting. Building for darwin is difficult for us. On Linux it is building fine as it is. Any clue on why this is happening?

**Comment by mrene** [comment]
> Thanks for reporting. Building for darwin is difficult for us. On Linux it is building fine as it is. Any clue on why this is happening?

It looks like a subtle filesystem casing issue, the modules have references to both `github.com/Microsoft` and `github.com/microsoft`. On linux they end up being distinct directories but because of darwin's default-case-insensitive filesystem, both of these directories end up being merged into one.

Linux:
```
❯ exa -TL1  ./result/github.com/{M,m}icrosoft/
./result/github.com/Microsoft
└── go-winio
./result/github.com/microsoft
└── go-mssqldb
```

Darwin:
```
❯ exa -TL1  ./result/github.com/Microsoft/
./result/github.com/Microsoft
├── go-mssqldb
└── go-winio
```

**Comment by superherointj** [comment]
@mrene Thanks for your helpful answer. Seems to be a case of fixing at upstream then.

**Comment by mrene** [comment]
Oh there seems to be an option to resolve that according to [the manual](https://nixos.org/manual/nixpkgs/stable/#ssec-language-go).

I get a different hash (`sha256-6JfxKzYAjSQ9JFuFGDUZ0ALS1D7f2LK3bP0Fbl2ivo0=`) but at it's consistent across platfoms.

```
proxyVendor: Fetches (go mod download) and proxies the vendor directory. This is useful if your code depends on c code and go mod tidy does not include the needed sources to build or if any dependency has case-insensitive conflicts which will produce platform dependant vendorHash checksums.
```

File: pkgs/development/tools/database/sqlcmd/default.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  src = fetchFromGitHub {
+    repo = "go-sqlcmd";
+    owner = "microsoft";
+    rev = "v${version}";
+    sha256 = "sha256-1CdZVh7pbupCNOE1ydgYEqX4rYo2LRddlNRNf0QjlN0=";
+  };
+
+  vendorSha256 = "sha256-/51qs7ACE9UmyE59Tsto+Xm240H6IUL8eNKfLgb5LUE=";
+
```
Comment (mrene) [inline pkgs/development/tools/database/sqlcmd/default.nix:20]:
```suggestion
  vendorSha256 = "sha256-6JfxKzYAjSQ9JFuFGDUZ0ALS1D7f2LK3bP0Fbl2ivo0=";
  proxyVendor = true;
```

**Comment by ratsclub** [comment]
What? I pushed a single file. Sorry for the notification, folks. Not sure what happened here.

**Pull request closed without merging**